hack: replace go-mod-prepare.sh with wrapper script

neersighted opened the pull request:
- follow-up to (and builds on) https://github.com/moby/moby/pull/44459

**- What I did**
To make the local build environment more correct and consistent, we should never leave an uncommitted go.mod in the tree; however, we need a go.mod for certain commands to work properly. Use a wrapper script to create and destroy the go.mod as needed instead of potentially changing tooling behavior by leaving it.

Also add a script to prevent go.mod from sneaking back in; as we are not a Go module, and cannot currently become one, we want to prevent new contributors from mistakenly trying to turn Moby into one.

**- How to verify it**
Manually, more or less.

**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://user-images.githubusercontent.com/1049222/202741498-2fd212af-9387-4ccb-83a7-69a543254392.png)

neersighted committed:
hack: replace go-mod-prepare.sh with wrapper script

To make the local build environment more correct and consistent, we
should never leave an uncommitted go.mod in the tree; however, we need a
go.mod for certain commands to work properly. Use a wrapper script to
create and destroy the go.mod as needed instead of potentially changing
tooling behavior by leaving it.

If a go.mod already exists, this script will warn and call the wrapped
command with GO111MODULE=on.

Signed-off-by: Bjorn Neergaard <bneergaard@mirantis.com>

corhere commented on hack/with-go-mod.sh:
```diff
@@ -0,0 +1,23 @@
+SCRIPTDIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+ROOTDIR="$(git -C "$SCRIPTDIR" rev-parse --show-toplevel)"
+
+set -x
+
+tee "${ROOTDIR}/go.mod" >&2 <<EOF
```
Clobbering an existing go.mod file without warning might not be ideal. For example, it could go really badly if (as I have done in the past) go.mod is a symlink to vendor.mod.
```suggestion
if [[ -e "${ROOTDIR}/go.mod" ]]; then
    >&2 echo "WARNING: go.mod already exists."
    exec "$@"
fi

set -x 

tee "${ROOTDIR}/go.mod" >&2 <<EOF
```

corhere commented on hack/vendor.sh:
```diff
@@ -7,18 +7,17 @@
 SCRIPTDIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
-"${SCRIPTDIR}"/go-mod-prepare.sh
 
 export GO111MODULE=on
```
Having to set `GO111MODULE=on` everywhere with-go-mod.sh is used is tedious, and I can't imagine a situation where one would need to use the wrapper script with any other `GO111MODULE` value. Move this line into with-go-mod.sh?

neersighted replied:
Good point!

corhere commented on .gitignore:
```diff
@@ -22,6 +22,3 @@ vendor/pkg/
 go-test-report.json
 profile.out
 junit-report.xml
-
-# top-level go.mod is not meant to be checked in
-/go.mod
```
Since it still is not meant to be committed, I don't think the .gitignore entry should be deleted unless there is some other automation to prevent go.mod from being merged from PRs. That is, I would approve this change iff "no go.mod" was enforced by CI, such as with a `hack/validate/` script.

neersighted replied:
I think a CI check makes more sense than a `.gitignore`, ignoring the go.mod is problematic because it may mask behavior changes from a go.mod file that appears from some other source.

neersighted committed:
hack: introduce validate/no-module

Moby is not a Go module; to prevent anyone from mistakenly trying to
convert it to one before we are ready, introduce a check (usable in CI
and locally) for a go.mod file.

This is preferable to trying to .gitignore the file as we can ensure
that a mistakenly created go.mod is surfaced by Git-based tooling and is
less likely to surprise a contributor.

Signed-off-by: Bjorn Neergaard <bneergaard@mirantis.com>

neersighted commented:
Pushed up a new version that adds a new validate step to go with the removal of go.mod from `.gitignore`.

thaJeztah commented:
@neersighted needs a rebase

neersighted commented:
This is stacked on #44459; do you want to land that and I'll rebase this on top of the merged branch?

thaJeztah commented:
The other one was merged; so this one can be rebased 👍

neersighted commented:
Rebased :+1:

thaJeztah commented:
Looks like shfmt isn't happy 🙈

thaJeztah commented on .gitignore:
```diff
@@ -13,8 +13,6 @@ thumbs.db
 .bashrc
 .editorconfig
 
-# top-level go.mod is not meant to be checked in
-/go.mod
```
Still _slightly_ on the fence if this could cause accidentally adding the go.mod (if the script failed at some point  etc), but I'll guess we'll see how it goes 🙈

corhere replied:
`hack/validate/no-module` added in this PR would catch those cases. Though it will need to be added to `hack/validate/default` to be run in CI?

neersighted replied:
I excluded it from `default` because some users may want to be able to have the tooling function with a go.mod present but not committed. It's run in CI as there is a file-based glob. However, we could add it to `all` (the same step where vendoring is run).

kolyshkin replied:
> It's run in CI as there is a file-based glob.

@neersighted I don't see it (neither in the current git HEAD, nor in the source tree during the state when this PR got merged.

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
@neersighted could you prepare a 🍒 ⛏️ for this one (and the previous PR)?
